Update to Zig 0.15.1

FalsePattern opened the pull request:
- Updated versions.nix using `nix run .#update-versions`
- Rewrote the `zig2nix` and `bundle/loader` source code to the new 0.15.1 IO api, and also fixed some future deprecations, like the `std.ArrayListUnmanaged` -> `std.ArrayList` rename
- Regenerated the template using `nix run .#update-templates`

I've done an initial validation using `zig test` and the `nix run .#test-all` task, the zig code compiles without issues, but the nix task failed on my system with `failed to setup an namespace, cannot continue`, so it will need a validation pass on CI.
I also tested it in a dev env, and it seems to be working properly aside from that.

Cloudef commented on src/zig2nix/src/cli.zig:
```diff
@@ -148,13 +152,18 @@ const ZigEnv = struct {
+    var proc_reader = proc.reader(&buf);
+    var tmp = try std.Io.Writer.Allocating.initCapacity(allocator, 1024);
+    defer tmp.deinit();
+    try @import("zon2json.zig").parse(allocator, &proc_reader.interface, &tmp.writer, error_writer, .{.file_name = "zig env"});
```
Could perhaps use std.zon instead. Also to support older zig versions, we need to support zig env that returns json as well.

FalsePattern commented:
I've added json env compat by just assuming that errors from the zon parser mean that it's a json-based `zig env`, which should be safe enough. I also added tests with a zig 0.14.1 env to the flake, they're under `test-0_14` (`test-0_14-all`, etc.)

I didn't touch the CI workflow files in the PR, because github would complain about it
